File: frontend/types/lesson.ts
export interface Lesson {
  id: string;
  student_id: string;
  tutor_id: string;
  date: string;
  start_time: string;
  end_time: string;
  price: number;
  comment?: string | null;
  status: LessonStatus;
  payment_status: PaymentStatus;
  student_name?: string | null;
  created_at: string;
}

export type LessonStatus = "scheduled" | "completed" | "cancelled";
export type PaymentStatus = "paid" | "unpaid";
export type RecurrenceType = "daily" | "weekly" | "biweekly" | "monthly";
export type CalendarViewMode = "day" | "week" | "month";

export interface LessonFilter {
  start_date?: string;
  end_date?: string;
  student_id?: string;
  status?: LessonStatus;
  page?: number;
  per_page?: number;
}

export interface LessonListResponse {
  lessons: Lesson[];
  total: number;
  page?: number;
  per_page?: number;
}

export interface LessonFormData {
  student_id: string;
  date: string;
  start_time: string;
  end_time: string;
  price: number;
  comment: string;
}

export interface RescheduleData {
  date: string;
  start_time: string;
  end_time: string;
}

export interface RecurringLessonData {
  recurrence_type: RecurrenceType;
  end_date?: string;
  total_occurrences?: number;
}

export interface Student {
  id: string;
  first_name: string;
  last_name: string;
  email?: string | null;
  phone?: string | null;
  subject?: string;
  lesson_price?: number;
  status: "active" | "paused" | "finished";
}

export interface StudentsResponse {
  students: Student[];
  total: number;
}

File: backend/app/schemas/lesson.py
from datetime import date, datetime, time
from typing import Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class LessonCreate(BaseModel):
    student_id: UUID
    date: date
    start_time: time
    end_time: time
    price: float | None = Field(None, gt=0)
    comment: str | None = None


class LessonUpdate(BaseModel):
    date: Optional[date] = None
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    price: float | None = Field(None, gt=0)
    comment: str | None = None
    status: Literal["scheduled", "completed", "cancelled"] | None = None


class LessonReschedule(BaseModel):
    date: date
    start_time: time
    end_time: time


class LessonPaymentUpdate(BaseModel):
    payment_status: Literal["paid", "unpaid"]


class LessonStatusUpdate(BaseModel):
    status: Literal["scheduled", "completed", "cancelled"]


class RecurrenceCreate(BaseModel):
    recurrence_type: Literal["daily", "weekly", "biweekly", "monthly"]
    end_date: date
    total_occurrences: int = Field(..., gt=1)


class LessonResponse(BaseModel):
    id: UUID
    student_id: UUID
    tutor_id: UUID
    date: date
    start_time: time
    end_time: time
    price: float
    comment: str | None
    status: str
    payment_status: str
    student_name: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class LessonListResponse(BaseModel):
    lessons: list[LessonResponse]
    total: int
    page: int
    per_page: int


class CalendarResponse(BaseModel):
    lessons: list[LessonResponse]
    start_date: date
    end_date: date
